app/src/main/java/org/mozilla/fenix/perf/ProfilerStopDialogFragment.kt

```diff
@@ -0,0 +1,236 @@
+                            }
+                        }
+                    ) {
+                        Text(text = stringResource(R.string.profiler_start_stop))
+                    }
+                }
+            }
+        }
+    }
+
+    @Composable
+    private fun UrlWarningCard(
+        viewStateObserver: MutableState<CardState>,
+        saveProfileTypeObserver: MutableState<String>
+    ) {
+        DialogueCard {
+            Column(modifier = Modifier.padding(8.dp)) {
+                Text(
+                    text = "Saving to URL warning",
+                    fontWeight = FontWeight.ExtraBold,
+                    fontSize = 20.sp,
+                    modifier = Modifier.padding(8.dp)
+                )
+                Spacer(modifier = Modifier.height(10.dp))
+                Text(
+                    text = "Saving to URL could potentially show private information such as " +
+                        "urls and screenshots. If you want to keep these" +
+                        "privates, save your profile as a file instead and head on over to profiler.firefox.com" +
+                        "where you can select what to share and not share.",
+                    fontWeight = FontWeight.Medium,
+                    fontSize = 15.sp,
+                    modifier = Modifier.padding(8.dp)
+                )
+                Spacer(modifier = Modifier.height(30.dp))
+                Row(
+                    horizontalArrangement = Arrangement.End,
+                    modifier = Modifier.fillMaxWidth()
+                ) {
+                    TextButton(
+                        onClick = {
+                            saveProfileTypeObserver.value = resources.getString(R.string.profiler_save_method_file)
+                            viewStateObserver.value = CardState.WaitForProfilerState
+                        }
+                    ) {
+                        Text("Save as File")
+                    }
+                    Spacer(modifier = Modifier.width(4.dp))
+
+                    TextButton(
+                        onClick = {
+                            viewStateObserver.value = CardState.WaitForProfilerState
+                        }
+                    ) {
+                        Text("Get Profile URL")
+                    }
+                }
+            }
+        }
+    }
+
+    private fun createFileSaverLauncher() {
+        profileSaveLauncher = registerForActivityResult(
+            ActivityResultContracts.StartActivityForResult()
+        ) { result ->
+            if (result.resultCode == Activity.RESULT_OK) {
+                val uri = result.data?.data
+                val jsonFile = context?.contentResolver?.openFileDescriptor(uri!!, "w")
+                val fileOutputStream = FileOutputStream(jsonFile!!.fileDescriptor)
+                fileOutputStream.write(profileArray)
+                fileOutputStream.close()
+                jsonFile.close()
```

Comment:
This should be safer. In particular, it handles closing streams if an exception is thrown (hopefully I got the syntax right):

```suggestion
                context?.contentResolver?.openFileDescriptor(uri!!, "w").use { jsonFile ->
                    FileOutputStream(jsonFile!!.fileDescriptor).use { fileOutputStream ->
                        fileOutputStream.write(profileArray)
                    }
                }
```